certbot/filelock.py

```diff
@@ -0,0 +1,324 @@
+    threads. It is based on the lock_file package by Martin Horcicka.
+    """
+
+    def __init__(self, path):
+        # type: (str) -> None
+        """
+        Create a lock file mechanism for Unix.
+        :param str path: the path to the lock file
+        """
+        self._path = path
+        self._fd = None  # type: Optional[int]
+
+    def acquire(self):
+        # type: () -> None
+        """Acquire the lock."""
+        while self._fd is None:
+            # Open the file
+            fd = os.open(self._path, os.O_CREAT | os.O_WRONLY, 0o600)
+            try:
+                self._try_lock(fd)
+                if self._lock_success(fd):
+                    self._fd = fd
+            finally:
+                # Close the file if it is not the required one
+                if self._fd is None:
+                    os.close(fd)
+
+    def _try_lock(self, fd):
+        # type: (int) -> None
+        """
+        Try to acquire the lock file without blocking.
+        :param int fd: file descriptor of the opened file to lock
+        """
+        try:
+            fcntl.lockf(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        except IOError as err:
+            if err.errno in (errno.EACCES, errno.EAGAIN):
+                logger.debug(
+                    "A lock on %s is held by another process.", self._path)
+                _raise_for_certbot_lock(self._path)
+            raise
+
+    def _lock_success(self, fd):
+        # type: (int) -> bool
+        """
+        Did we successfully grab the lock?
+
+        Because this class deletes the locked file when the lock is
+        released, it is possible another process removed and recreated
+        the file between us opening the file and acquiring the lock.
+
+        :param int fd: file descriptor of the opened file to lock
+        :returns: True if the lock was successfully acquired
+        :rtype: bool
+        """
+        try:
+            stat1 = os.stat(self._path)
+        except OSError as err:
+            if err.errno == errno.ENOENT:
+                return False
+            raise
+
+        stat2 = os.fstat(fd)
+        # If our locked file descriptor and the file on disk refer to
+        # the same device and inode, they're the same file.
+        return stat1.st_dev == stat2.st_dev and stat1.st_ino == stat2.st_ino
+
+    def release(self):
+        # type: () -> None
+        """Remove, close, and release the lock file."""
+        # It is important the lock file is removed before it's released,
+        # otherwise:
+        #
+        # process A: open lock file
+        # process B: release lock file
+        # process A: lock file
+        # process A: check device and inode
+        # process B: delete file
+        # process C: open and lock a different file at the same path
+        try:
+            os.remove(self._path)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+        finally:
+            if self._fd:
```

Comment:
This is needed to make mypy happy. Otherwise, as self._fd can be None or an int, it as the type Optional[int], while os.close expect an int. Good occasion in fact, to see `mypy` in action to find potential bugs.